src/main/scala/com/actian/spark_vector/colbuffer/ColumnBuffer.scala

```diff
@@ -22,89 +22,158 @@ import com.actian.spark_vector.colbuffer.singles._
+private[colbuffer] sealed abstract class RWColumnBuffer(col: ColumnBuffer[_, _])
+  extends Serializable {
+  protected final val NullMarker = 1: Byte
+  protected final val NonNullMarker = 0: Byte
+
+  val alignSize = col.alignSize
+  val nullable = col.nullable
+  val maxValueCount = col.maxValueCount
+
+  def position(): Int
+
+  def clear(): Unit
+}
 
-  protected def put(source: T, buffer: ByteBuffer): Unit
+class WriteColumnBuffer[@specialized T: ClassTag](col: ColumnBuffer[T, _]) extends RWColumnBuffer(col) {
+  val values = ByteBuffer.allocateDirect(col.maxValueCount * col.valueWidth).order(ByteOrder.nativeOrder())
+  private val nullValue = Array.fill[Byte](col.alignSize)(0: Byte)
+  lazy val markers = ByteBuffer.allocateDirect(col.maxValueCount).order(ByteOrder.nativeOrder())
+
+  /** Make valueType visible outside */
+  val valueType = col.valueTypeIn
 
   def put(source: T): Unit = {
-    put(source, values)
-    if (nullable) {
+    col.put(source, values)
+    if (col.nullable) {
       markers.put(NonNullMarker)
     }
   }
 
-  @throws(classOf[IllegalArgumentException])
   def putNull(): Unit = {
-    if (!nullable) {
-      throw new IllegalArgumentException(s"Cannot store NULL values in non-nullable '${name}' column.")
-    }
+    if (!col.nullable) throw new IllegalStateException(s"Cannot store NULLs in non-nullable column '${col.name}'.")
     markers.put(NullMarker)
     values.put(nullValue)
   }
 
-  def size: Int = {
+  override def position(): Int = {
     var ret = values.position()
-    if (nullable) {
+    if (col.nullable) {
       ret += markers.position()
     }
     ret
   }
 
   def flip(): Unit = {
     values.flip()
-    if (nullable) {
+    if (col.nullable) {
       markers.flip()
     }
   }
 
-  def clear(): Unit = {
+  override def clear(): Unit = {
     values.clear()
-    if (nullable) {
+    if (col.nullable) {
       markers.clear()
     }
   }
 }
 
-/** Trait to be used when implementing a companion object for a typed ColumnBuffer
+class ReadColumnBuffer[@specialized T: ClassTag](col: ColumnBuffer[_, T]) extends RWColumnBuffer(col) {
+  private val values = Array.ofDim[T](col.maxValueCount)
+  private val isNullValue = Array.ofDim[Boolean](col.maxValueCount)
+  private lazy val markers = Array.ofDim[Byte](col.maxValueCount)
+
+  private var leftPos = 0
+  private var rightPos = 0
+
+  /** Make valueType visible outside */
+  val valueType = col.valueTypeOut
+
+  private def isEmpty = leftPos >= rightPos
+
+  def fill(source: ByteBuffer, n: Int): Unit = {
+    val nLimit = Math.min(n, maxValueCount)
+    if (col.nullable) {
+      source.get(markers, 0, nLimit)
+    }
+    var pad = padding(IntSize /* messageLength, not incl. in source position */ + source.position, col.alignSize)
+    source.position(source.position + pad)
+    while (rightPos < nLimit) {
+      isNullValue(rightPos) = col.nullable.ifThenElse(markers(rightPos) == NullMarker, false)
```

Comment:
shouldn't use implicit classes here (they will allocate an object for each tuple here).